javalib/src/main/scala/java/util/ArrayDeque.scala

```diff
@@ -0,0 +1,193 @@
+      status += 1L
+      true
+    }
+  }
+
+  def offerLast(e: E): Boolean = {
+    if (e == null) {
+      throw new NullPointerException()
+    } else {
+      inner += e
+      status += 1L
+      true
+    }
+  }
+
+  def removeFirst(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollFirst()
+  }
+
+  def removeLast(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      pollLast()
+  }
+
+  def pollFirst(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else {
+      val res = inner.remove(0)
+      status += 1L
+      res
+    }
+  }
+
+  def pollLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.pop()
+  }
+
+  def getFirst(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      peekFirst()
+  }
+
+  def getLast(): E = {
+    if (inner.isEmpty)
+      throw new NoSuchElementException()
+    else
+      peekLast()
+  }
+
+  def peekFirst(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.head
+  }
+
+  def peekLast(): E = {
+    if (inner.isEmpty) null.asInstanceOf[E]
+    else inner.last
+  }
+
+  def removeFirstOccurrence(o: Any): Boolean = {
+    val index = inner.indexWhere(_ === o)
+    if (index >= 0) {
+      val res = inner.remove(index) != null
+      if (res) status += 1L
+      res
+    } else
+      false
+  }
+
+  def removeLastOccurrence(o: Any): Boolean = {
+    val index = inner.lastIndexWhere(_ === o)
+    if (index >= 0) {
+      val res = inner.remove(index) != null
+      if (res) status += 1L
+      res
+    } else
+      false
+  }
+
+  override def add(e: E): Boolean = {
+    addLast(e)
+    true
+  }
+
+  def offer(e: E): Boolean = offerLast(e)
+
+  def remove(): E = removeFirst()
+
+  def poll(): E = pollFirst()
+
+  def element(): E = getFirst()
+
+  def peek(): E = peekFirst()
+
+  def push(e: E): Unit = addFirst(e)
+
+  def pop(): E = removeFirst()
+
+  def size(): Int = inner.size
+
+  override def isEmpty(): Boolean = inner.isEmpty
+
+  private var status = 0L
+
+  private def _failFastIterator(iter: scala.collection.Iterator[E]) = {
+    new Iterator[E] {
+      private def checkStatus() =
+        if (self.status != actualStatus)
+          throw new ConcurrentModificationException()
+
+      private val actualStatus = self.status
+
+      private var index: Int = -1
+
+      def hasNext(): Boolean = {
+        checkStatus()
+        iter.hasNext
+      }
+
+      def next(): E = {
+        checkStatus()
+        index += 1
+        iter.next()
+      }
+
+      def remove(): Unit = {
+        checkStatus()
+        if (index <= 0) {
+          throw new IllegalStateException()
+        } else {
+          inner.remove(index)
+        }
+      }
+    }
+  }
+
+  def iterator(): Iterator[E] =
+    _failFastIterator(inner.iterator)
+
+  def descendingIterator(): Iterator[E] =
+    _failFastIterator(inner.reverse.iterator)
```

Comment:
This is not efficient. I would suggest implementing `_failFastIterator` based on direct access of `inner` using the index and abstract on the way the `index` is updated.